pkg/types/codec.go

```diff
@@ -29,6 +29,65 @@ type Decoder interface {
 	GetMaxDecodingSize(ctx context.Context, n int, itemType string) (int, error)
 }
 
+/*
+Codec is an interface that provides encoding and decoding functionality for a specific type identified by a name.
+Because there are many types that a ContractReader or ContractWriter can either accept or return, all encoding
+instructions provided by the codec are based on the type name.
+
+Starting from the lowest level, take for instance a big.Int encoder where we want the output to be big endian binary
+encoded.
+
+	typeCodec, _ := binary.BigEndian().BigInt(32, true)
+
+This allows us to encode and decode big.Int values with big endian encoding using the `encodings.TypeCodec` interface.
+
+	encodedBytes := []byte{}
+
+	originalValue := big.NewInt(42)
+	encodedBytes, _ = typeCodec.Encode(originalValue, encodedBytes) // new encoded bytes are appended to existing
+
+	value, _, _ := typeCodec.Decode(encodedBytes, value)
+
+The additional encodings.TypeCodec methods such as `GetType() reflect.Type` allow composition. This is useful for
```

Comment:
Might need square brackets on types refs to other packages?
```suggestion
The additional [encodings.TypeCodec] methods such as `GetType() reflect.Type` allow composition. This is useful for
```